In [64]:
import pandas as pd
import json
import xml.etree.ElementTree as ET

In [65]:
# https://api.greatschools.org/schools/CA/1?key=ntrebpcrjjzxtlq8oxwlpaxx

In [15]:
import urllib.request
import urllib.parse

In [88]:
text_file = open("apikey.txt", "r")
apikey = text_file.read()

str

In [25]:
url='https://api.greatschools.org/schools/CA/1?key='+apikey
f = urllib.request.urlopen(url)
xmlstr = f.read().decode('utf-8')

In [34]:
xmlobj = ET.fromstring(xmlobj)

In [63]:
xmlobj.attrib['type']

KeyError: 'type'

In [ ]:
elem = Element("tag")
elem.attrib["first"] = "1"
elem.attrib["second"] = "2"

In [55]:
for child in xmlobj:
    print(child)

<Element 'gsId' at 0x10b512f48>
<Element 'name' at 0x10b512f98>
<Element 'type' at 0x10b523048>
<Element 'gradeRange' at 0x10b523098>
<Element 'enrollment' at 0x10b5230e8>
<Element 'gsRating' at 0x10b523138>
<Element 'city' at 0x10b523188>
<Element 'state' at 0x10b5231d8>
<Element 'districtId' at 0x10b523228>
<Element 'district' at 0x10b523278>
<Element 'districtNCESId' at 0x10b5232c8>
<Element 'address' at 0x10b523318>
<Element 'phone' at 0x10b523368>
<Element 'fax' at 0x10b5233b8>
<Element 'website' at 0x10b523408>
<Element 'ncesId' at 0x10b523458>
<Element 'lat' at 0x10b5234a8>
<Element 'lon' at 0x10b5234f8>
<Element 'overviewLink' at 0x10b523548>
<Element 'ratingsLink' at 0x10b523598>
<Element 'reviewsLink' at 0x10b5235e8>
<Element 'parentReviews' at 0x10b523638>
<Element 'schoolStatsLink' at 0x10b523688>


In [59]:
xmlobj[4].text

'1853'

In [60]:
url="https://api.greatschools.org/schools/NY/New-York-City?key="+apikey
f = urllib.request.urlopen(url)
xmlstr = f.read().decode('utf-8')
xmlobj2 = ET.fromstring(xmlstr)

In [62]:
print(xmlstr)

<?xml version="1.0" encoding="UTF-8" standalone="yes"?><schools/>


In [66]:
url="https://api.greatschools.org/cities/CA/San-Francisco?key="+apikey
f = urllib.request.urlopen(url)
xmlstr = f.read().decode('utf-8')
print(xmlstr)

<?xml version="1.0" encoding="UTF-8" standalone="yes"?><city><name>San Francisco</name><rating>7</rating><totalSchools>274</totalSchools><elementarySchools>195</elementarySchools><middleSchools>143</middleSchools><highSchools>84</highSchools><publicSchools>109</publicSchools><charterSchools>18</charterSchools><privateSchools>147</privateSchools></city>


In [68]:
wootest = xml.dom.minidom.parseString(xmlstr)
pretty_xml_as_string = wootest.toprettyxml()

In [70]:
print(pretty_xml_as_string)

<?xml version="1.0" ?>
<city>
	<name>San Francisco</name>
	<rating>7</rating>
	<totalSchools>274</totalSchools>
	<elementarySchools>195</elementarySchools>
	<middleSchools>143</middleSchools>
	<highSchools>84</highSchools>
	<publicSchools>109</publicSchools>
	<charterSchools>18</charterSchools>
	<privateSchools>147</privateSchools>
</city>



In [71]:
def apipull(url):
    f = urllib.request.urlopen(url)
    xmlstr = f.read().decode("utf-8")
    return xmlstr

In [78]:
import xml.dom.minidom
def printxml(xmlstr):
    temp = xml.dom.minidom.parseString(xmlstr)
    print(temp.toprettyxml())

<?xml version="1.0" encoding="UTF-8" standalone="yes"?><city><name>New York</name><rating>6</rating><totalSchools>625</totalSchools><elementarySchools>386</elementarySchools><middleSchools>318</middleSchools><highSchools>265</highSchools><publicSchools>327</publicSchools><charterSchools>66</charterSchools><privateSchools>232</privateSchools></city>


In [ ]:
testnyc = apipull("https://api.greatschools.org/cities/NY/New-York?key="+apikey)
print(testnyc)

In [79]:
printxml(testnyc)

<?xml version="1.0" ?>
<city>
	<name>New York</name>
	<rating>6</rating>
	<totalSchools>625</totalSchools>
	<elementarySchools>386</elementarySchools>
	<middleSchools>318</middleSchools>
	<highSchools>265</highSchools>
	<publicSchools>327</publicSchools>
	<charterSchools>66</charterSchools>
	<privateSchools>232</privateSchools>
</city>



So it looks like there are 625 schools in New York City, only 52% of which are public. Wow

Note to self, api does not include preschools while search does

In [ ]:
testboston = apipull("https://api.greatschools.org/cities/MA/Boston?key="+apikey)

In [82]:
printxml(testboston)

<?xml version="1.0" ?>
<city>
	<name>Boston</name>
	<rating>4</rating>
	<totalSchools>199</totalSchools>
	<elementarySchools>136</elementarySchools>
	<middleSchools>93</middleSchools>
	<highSchools>64</highSchools>
	<publicSchools>118</publicSchools>
	<charterSchools>25</charterSchools>
	<privateSchools>56</privateSchools>
</city>



In [91]:
allnycschools = apipull("https://api.greatschools.org/schools/NY/New-York?key="+apikey+"&limit=-1")
printxml(allnycschools)    

<?xml version="1.0" ?>
<schools>
	<school>
		<gsId>8289</gsId>
		<name>47 the American Sign Language and English Lower School</name>
		<type>public</type>
		<gradeRange>PK-8</gradeRange>
		<enrollment>195</enrollment>
		<gsRating>7</gsRating>
		<parentRating>4</parentRating>
		<city>New York</city>
		<state>NY</state>
		<districtId>806</districtId>
		<district>Nyc Geog District # 2 - Manhattan</district>
		<districtNCESId>3600077</districtNCESId>
		<address>223 East 23rd Street, 
New York, NY  10010</address>
		<phone>(914) 326-6609</phone>
		<fax/>
		<website/>
		<ncesId>360007705767</ncesId>
		<lat>40.73862</lat>
		<lon>-73.98154</lon>
		<overviewLink>https://www.greatschools.org/new-york/new-york-city/8289-47-The-American-Sign-Language-And-English-Lower-School/?s_cid=gsapi</overviewLink>
		<ratingsLink>https://www.greatschools.org/school/rating.page?state=NY&amp;id=8289&amp;s_cid=gsapi</ratingsLink>
		<reviewsLink>https://www.greatschools.org/school/parentReviews.page?state=NY&amp;i